File: clients/typescript/src/httpsign.test.ts
import assert from "node:assert/strict";
import { createPublicKey, verify as cryptoVerify } from "node:crypto";
import { describe, it } from "node:test";
import {
	REQUIRED_COMPONENTS,
	contentDigestSha256,
	generateEd25519Pem,
	pathAndQuery,
	signHeaders,
} from "./httpsign.ts";

describe("httpsign", () => {
	it("pathAndQuery handles absolute URLs and paths", () => {
		assert.deepEqual(pathAndQuery("http://localhost:8080/v1/chat/completions"), {
			path: "/v1/chat/completions",
			query: "?",
		});
		assert.deepEqual(pathAndQuery("/v1/models?foo=1"), {
			path: "/v1/models",
			query: "?foo=1",
		});
	});

	it("contentDigestSha256 shape", () => {
		const d = contentDigestSha256("hello");
		assert.ok(d.startsWith("sha-256=:"));
		assert.ok(d.endsWith(":"));
	});

	it("signHeaders produces a verifiable signature", () => {
		const { privateKeyPem, publicKeyPem } = generateEd25519Pem();
		const body = '{"ping":true}';
		const headers = signHeaders({
			method: "POST",
			url: "http://localhost:8080/v1/chat/completions",
			body,
			privateKeyPem,
			keyId: "local-signer",
			nonce: "n1",
			created: 1_700_000_000,
		});
		assert.ok(headers["Signature-Input"]?.startsWith("sig1="));
		assert.ok(headers.Signature?.startsWith("sig1=:"));
		for (const c of REQUIRED_COMPONENTS) {
			assert.ok(headers["Signature-Input"]?.includes(c));
		}

		const digest = headers["Content-Digest"]!;
		const sigParams = headers["Signature-Input"]!.slice("sig1=".length);
		const sigBase = [
			`"@method": POST`,
			`"@path": /v1/chat/completions`,
			`"@query": ?`,
			`"content-digest": ${digest}`,
			`"@signature-params": ${sigParams}`,
		].join("\n");
		const raw = Buffer.from(
			headers.Signature!.slice("sig1=:".length, -1),
			"base64",
		);
		const ok = cryptoVerify(
			null,
			Buffer.from(sigBase, "utf8"),
			createPublicKey(publicKeyPem),
			raw,
		);
		assert.equal(ok, true);
	});
});

File: clients/java/src/test/java/ai/afi/platform/HttpSignTest.java
package ai.afi.platform;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.charset.StandardCharsets;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.Signature;
import java.util.Base64;
import java.util.Map;
import org.junit.jupiter.api.Test;

class HttpSignTest {

  private static byte[] pemPrivate(KeyPair pair) {
    String b64 = Base64.getMimeEncoder(64, new byte[] {'\n'}).encodeToString(pair.getPrivate().getEncoded());
    return ("-----BEGIN PRIVATE KEY-----\n" + b64 + "\n-----END PRIVATE KEY-----\n")
        .getBytes(StandardCharsets.US_ASCII);
  }

  private static KeyPair ed25519() throws Exception {
    return KeyPairGenerator.getInstance("Ed25519").generateKeyPair();
  }

  @Test
  void contentDigestShape() {
    String d = HttpSign.contentDigestSha256("hello".getBytes(StandardCharsets.UTF_8));
    assertTrue(d.startsWith("sha-256=:"));
    assertTrue(d.endsWith(":"));
  }

  @Test
  void signHeadersRoundtripVerifyBase() throws Exception {
    KeyPair pair = ed25519();
    byte[] pem = pemPrivate(pair);
    byte[] body = "{\"ping\":true}".getBytes(StandardCharsets.UTF_8);
    Map<String, String> headers =
        HttpSign.signHeaders(
            "POST",
            "http://localhost:8080/v1/chat/completions",
            body,
            pem,
            "local-signer",
            "n1",
            1_700_000_000L);
    assertTrue(headers.containsKey("Content-Digest"));
    assertTrue(headers.get("Signature-Input").startsWith("sig1="));
    assertTrue(headers.get("Signature").startsWith("sig1=:"));
    for (String c : HttpSign.REQUIRED_COMPONENTS) {
      assertTrue(headers.get("Signature-Input").contains(c));
    }

    String digest = headers.get("Content-Digest");
    String sigParams = headers.get("Signature-Input").substring("sig1=".length());
    String sigBase =
        String.join(
            "\n",
            "\"@method\": POST",
            "\"@path\": /v1/chat/completions",
            "\"@query\": ?",
            "\"content-digest\": " + digest,
            "\"@signature-params\": " + sigParams);
    String sigB64 =
        headers
            .get("Signature")
            .substring("sig1=:".length(), headers.get("Signature").length() - 1);
    byte[] raw = Base64.getDecoder().decode(sigB64);
    Signature verifier = Signature.getInstance("Ed25519");
    verifier.initVerify(pair.getPublic());
    verifier.update(sigBase.getBytes(StandardCharsets.UTF_8));
    assertTrue(verifier.verify(raw));
  }

  @Test
  void signHeadersWithQuery() throws Exception {
    KeyPair pair = ed25519();
    byte[] pem = pemPrivate(pair);
    Map<String, String> headers =
        HttpSign.signHeaders("GET", "/v1/models?foo=1", new byte[0], pem, "kid", "n", 1L);
    assertTrue(headers.get("Signature-Input").contains("@query"));
    String digest = headers.get("Content-Digest");
    String sigParams = headers.get("Signature-Input").substring("sig1=".length());
    String sigBase =
        String.join(
            "\n",
            "\"@method\": GET",
            "\"@path\": /v1/models",
            "\"@query\": ?foo=1",
            "\"content-digest\": " + digest,
            "\"@signature-params\": " + sigParams);
    String sigB64 =
        headers
            .get("Signature")
            .substring("sig1=:".length(), headers.get("Signature").length() - 1);
    Signature verifier = Signature.getInstance("Ed25519");
    verifier.initVerify(pair.getPublic());
    verifier.update(sigBase.getBytes(StandardCharsets.UTF_8));
    assertTrue(verifier.verify(Base64.getDecoder().decode(sigB64)));
  }

  @Test
  void pathAndQuery() {
    HttpSign.PathAndQuery abs = HttpSign.pathAndQuery("http://x/v1/chat?a=1");
    assertEquals("/v1/chat", abs.path());
    assertEquals("?a=1", abs.query());
    HttpSign.PathAndQuery empty = HttpSign.pathAndQuery("/v1/models");
    assertEquals("/v1/models", empty.path());
    assertEquals("?", empty.query());
  }

  @Test
  void loadRoundtripKeyFactory() throws Exception {
    KeyPair pair = ed25519();
    byte[] pem = pemPrivate(pair);
    var priv = HttpSign.loadPrivateKey(pem);
    assertTrue(priv.getAlgorithm().equals("Ed25519") || priv.getAlgorithm().equals("EdDSA"));
  }
}
